#24: Fix timeout on GitHub actions (#25)
exasol/small-json-files-test-fixture: README.md

Creating 1,000,000 S3 objects costs about $5. Compared to that storage is cheap, since the files are so small.

## Architecture

> Creating and deleting many small files in S3 requires many HTTP requests. To reduce network latency for single requests and therefore speed up execution time we execute the requests for creating and deleting in an AWS Lambda. That's why we split the code into the client library written in Java and lambda code written in JavaScript.

Reviewer: ```suggestion
Creating and deleting many small files in S3 requires many HTTP requests. To reduce network latency for each request and by that speed up the execution the requests for creating and deleting are executed in an AWS Lambda. The controlling code is therefore split into a client library written in Java and lambda code written in JavaScript.
```

### Deleting Files

> The time required for deleting files is dominated by the time to list all files in S3 as listing only returns 1,000 files at a time. That's why the Java code calls the lambda with request action type `delete`. The lambda lists all files, collects the file names and once 50,000 files are found, invokes another instance of itself with request action type `deleteList`, passing the files. This way listing and deleting can run in parallel.

Reviewer: ```suggestion
The time required for deleting files is dominated by the time to list all files in S3 as listing only returns 1,000 files at a time. The Java code therefore needs to call the parent instance of the lambda with request action type `delete` multiple times. 

The parent lambda 
* lists all files
* collects the file names and 
* once 50,000 files are found, invokes a child instance of itself with request action type `deleteList`, passing the files. 

This way listing and deleting can run in parallel.
```

> Please note that using this approach we can only delete around 1,100,000 files before hitting the lambda execution limit of 15 minutes.

Reviewer: ```suggestion
Please note that using this approach we can only delete around 1,100,000 files before hitting the execution limit of 15 minutes for the root instance of the lambda function.
```

We need to consider the following constraints when creating 1,000,000 files on S3:

* AWS Lambdas have a maximum runtime of 15 minutes, so we can't create too many files in a single lambda. 20,000 files per lambda seems to be OK.
* When invoking too many lambdas in parallel (around 100) they will send too many requests in a short time to S3. Some of these requests will then fail with error `SlowDown: Please reduce your request rate`. 50 parallel invocations (i.e. 20,000 files per lambda) seem to be OK.
  * The recommended solution to use a random hash in the file name to allow S3 partitioning didn't work.
* We want to invoke Lambdas synchronously so that we know when they are done. In order to avoid HTTP request timeouts we increased the read timeout for the S3 library's HTTP client to more than 15 minutes. This works, but recently these long running HTTP request (20,000 files per lambda) started to fail with a read timeout on GitHub Actions, see issue [#24](https://github.com/exasol/small-json-files-test-fixture/issues/24). Reducing the number of files per lambda to 5,000 reduces the runtime sufficiently to avoid these timeouts.

> To solve these constraints we use a fixed thread pool for synchronously invoking the lambdas. This way only a limited number of lambdas is running concurrently. Experiments showed that a concurrency of 75 works well.

Reviewer: ```suggestion
* AWS Lambdas have a maximum runtime of 15 minutes, so we can't create all files in a single lambda. 
* The number of concurrent requests to S3 is limited. When invoking too many lambdas in parallel, S3 causes a failure with error  `SlowDown: Please reduce your request rate`. 
* The recommended solution to use a random hash in the file name to allow S3 partitioning didn't work.

SJFTF uses the following approach: 
* The Java client invokes the Lambdas synchronously so that he knows when they are done. 
* Additionally the client increases the read timeout for the S3 library's HTTP client to more than 15 minutes. 
* Each lambda creates only 5 000 files to avoid a timeout for the HTTP request observed by the GitHub Action, see issue [#24](https://github.com/exasol/small-json-files-test-fixture/issues/24). 
The Java client uses a thread pool to limit the max. number of concurrent lambdas to 75.
```
